# Customer Churn Exploratory Data Analysis

## 1. Data Understanding

## Project Objective

The objective of this project is to analyze customer-level data to understand
patterns associated with customer churn.

The analysis will focus on customer demographics, tenure, subscribed services,
contract characteristics, payment methods, and customer charges.

The goal is to identify customer segments and factors associated with higher
churn and translate the findings into actionable customer retention insights.

In [1]:
import pandas as pd

### Load the Dataset

The raw customer dataset is loaded into a Pandas DataFrame so
that its structure and contents can be inspected.

In [2]:
df = pd.read_csv("../data/raw/customer_churn.csv")

In [3]:
df.shape

(7262, 22)

In [4]:
df.head()

,CustomerID,Gender,Age,SeniorCitizen,Partner,Dependents,Tenure,PhoneService,MultipleLines,InternetService,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,C06889,Male,54.0,0,No,No,37,Yes,No,Fiber optic,...,No,No,Yes,No,Two year,No,Bank transfer,67.75,2661.46,No
1,C00415,Female,46.0,0,Yes,No,16,No,No phone service,DSL,...,No,No,No,No,Month-to-month,Yes,Electronic check,52.25,978.53,No
2,C03600,Female,34.0,0,No,No,0,Yes,No,DSL,...,No,Yes,No,No,One year,No,Electronic check,53.07,41.96,No
3,C01028,Female,52.0,0,No,No,23,Yes,Yes,Fiber optic,...,Yes,No,No,No,Month-to-month,Yes,Bank transfer,71.13,NaN,No
4,C04489,Female,29.0,0,No,No,16,Yes,No,No,...,No internet service,No internet service,No internet service,No internet service,Month-to-month,Yes,Mailed check,33.48,597.84,No


In [5]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 7262 entries, 0 to 7261
Data columns (total 22 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   CustomerID        7262 non-null   str    
 1   Gender            7262 non-null   str    
 2   Age               7175 non-null   float64
 3   SeniorCitizen     7262 non-null   int64  
 4   Partner           7262 non-null   str    
 5   Dependents        7262 non-null   str    
 6   Tenure            7262 non-null   int64  
 7   PhoneService      7262 non-null   str    
 8   MultipleLines     7262 non-null   str    
 9   InternetService   7262 non-null   str    
 10  OnlineSecurity    7262 non-null   str    
 11  OnlineBackup      7262 non-null   str    
 12  DeviceProtection  7262 non-null   str    
 13  TechSupport       7160 non-null   str    
 14  StreamingTV       7262 non-null   str    
 15  StreamingMovies   7262 non-null   str    
 16  Contract          7262 non-null   str    
 17  Paperl

In [6]:
df.columns

Index(['CustomerID', 'Gender', 'Age', 'SeniorCitizen', 'Partner', 'Dependents',
       'Tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='str')

In [7]:
df.describe()

,Age,SeniorCitizen,Tenure,MonthlyCharges,TotalCharges
count,7175.000000,7262.000000,7262.000000,7204.000000,7175.000000
mean,43.065784,0.067612,28.910218,56.925479,1714.135072
std,14.166101,0.251097,19.110820,17.880380,1247.965256
min,18.000000,0.000000,0.000000,18.000000,0.350000
25%,33.000000,0.000000,14.000000,45.832500,742.080000
50%,43.000000,0.000000,28.000000,59.785000,1509.560000
75%,53.000000,0.000000,43.000000,69.930000,2488.585000
max,80.000000,1.000000,72.000000,101.200000,7084.300000


In [9]:
df.describe(include="object")

C:\Users\ASUS  VIVOBOOK\AppData\Local\Temp\ipykernel_12892\702825166.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  df.describe(include="object")


,CustomerID,Gender,Partner,Dependents,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,Churn
count,7262,7262,7262,7262,7262,7262,7262,7262,7262,7262,7160,7262,7262,7262,7262,7174,7262
unique,7250,2,4,4,4,3,3,3,3,3,3,3,3,3,4,4,2
top,C06068,Female,No,No,Yes,No,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Electronic check,No
freq,2,3650,4091,5610,6536,3656,3486,3976,3610,3721,3983,3494,3481,3968,4272,2740,5422


## Missing Value Assessment

In [16]:
df.isnull().sum()

# This tells you the number of missing values per column.

CustomerID            0
Gender                0
Age                  87
SeniorCitizen         0
Partner               0
Dependents            0
Tenure                0
PhoneService          0
MultipleLines         0
InternetService       0
OnlineSecurity        0
OnlineBackup          0
DeviceProtection      0
TechSupport         102
StreamingTV           0
StreamingMovies       0
Contract              0
PaperlessBilling      0
PaymentMethod        88
MonthlyCharges       58
TotalCharges         87
Churn                 0
dtype: int64

In [12]:
missing_count = df.isnull().sum()

missing_percentage = (missing_count / len(df)) * 100

missing_summary = pd.DataFrame({
    "MissingCount": missing_count,
    "MissingPercentage": missing_percentage
})

missing_summary.sort_values(
    "MissingPercentage",
    ascending=False
)

,MissingCount,MissingPercentage
TechSupport,102,1.404572
PaymentMethod,88,1.211787
Age,87,1.198017
TotalCharges,87,1.198017
MonthlyCharges,58,0.798678
CustomerID,0,0.000000
Dependents,0,0.000000
Partner,0,0.000000
SeniorCitizen,0,0.000000
Gender,0,0.000000


### Check Duplicate Customers

Since the analysis is at customer level, I first checked whether CustomerID uniquely identifies each record. This is important because duplicate customer records could bias churn rates and segment-level comparisons.

In [13]:
df["CustomerID"].nunique()

7250

In [14]:
len(df)

7262

In [15]:
df["CustomerID"].duplicated().sum()

np.int64(12)

### Understand Categorical Variables

In [17]:
df["Gender"].value_counts()

Gender
Female    3650
Male      3612
Name: count, dtype: int64

In [19]:
# But instead of manually checking every column one by one, we can identify categorical columns.

categorical_columns = df.select_dtypes(include="object").columns

for col in categorical_columns:
    print(f"\n--- {col} ---")
    print(df[col].value_counts(dropna=False))


--- CustomerID ---
CustomerID
C06068    2
C07014    2
C06999    2
C05251    2
C06422    2
         ..
C02246    1
C02108    1
C04646    1
C03002    1
C05735    1
Name: count, Length: 7250, dtype: int64

--- Gender ---
Gender
Female    3650
Male      3612
Name: count, dtype: int64

--- Partner ---
Partner
No     4091
Yes    3142
NO       16
yes      13
Name: count, dtype: int64

--- Dependents ---
Dependents
No     5610
Yes    1623
NO       21
yes       8
Name: count, dtype: int64

--- PhoneService ---
PhoneService
Yes    6536
No      697
yes      26
NO        3
Name: count, dtype: int64

--- MultipleLines ---
MultipleLines
No                  3656
Yes                 2906
No phone service     700
Name: count, dtype: int64

--- InternetService ---
InternetService
Fiber optic    3486
DSL            2473
No             1303
Name: count, dtype: int64

--- OnlineSecurity ---
OnlineSecurity
No                     3976
Yes                    1983
No internet service    1303
Name: count, dtyp

C:\Users\ASUS  VIVOBOOK\AppData\Local\Temp\ipykernel_12892\2446813981.py:3: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_columns = df.select_dtypes(include="object").columns


### Identify Numerical Variables

In [20]:
numerical_columns = df.select_dtypes(
    include=["int64", "float64"]
).columns

numerical_columns

Index(['Age', 'SeniorCitizen', 'Tenure', 'MonthlyCharges', 'TotalCharges'], dtype='str')

In [21]:
df[numerical_columns].describe()

,Age,SeniorCitizen,Tenure,MonthlyCharges,TotalCharges
count,7175.000000,7262.000000,7262.000000,7204.000000,7175.000000
mean,43.065784,0.067612,28.910218,56.925479,1714.135072
std,14.166101,0.251097,19.110820,17.880380,1247.965256
min,18.000000,0.000000,0.000000,18.000000,0.350000
25%,33.000000,0.000000,14.000000,45.832500,742.080000
50%,43.000000,0.000000,28.000000,59.785000,1509.560000
75%,53.000000,0.000000,43.000000,69.930000,2488.585000
max,80.000000,1.000000,72.000000,101.200000,7084.300000


# One Important Thing About SeniorCitizen

SeniorCitizen → int64

Technically, it's numerical.

But conceptually, it is a categorical/binary variable.

### Understand the Churn Variable

In [22]:
df["Churn"].value_counts()

Churn
No     5422
Yes    1840
Name: count, dtype: int64

In [ ]:
# Gives you Percentage.
df["Churn"].value_counts(normalize=True) * 100

Churn
No     74.662627
Yes    25.337373
Name: proportion, dtype: float64

## Initial Observations

- The dataset contains 7,262 customer records and 22 columns.
- CustomerID is expected to represent the unique customer identifier and will
  be checked for duplicates.
- The dataset contains both numerical and categorical variables.
- Missing values are present in several columns and will be investigated before
  deciding on an appropriate treatment.
- Churn is the primary business outcome variable.
- Numerical variables such as Age, Tenure, MonthlyCharges and TotalCharges
  require distribution analysis before further processing.
- Categorical variables will be checked for unique values and inconsistencies.

# Investigate duplicate CustomerIDs

In [24]:
df[df["CustomerID"].duplicated(keep=False)].sort_values("CustomerID")

,CustomerID,Gender,Age,SeniorCitizen,Partner,Dependents,Tenure,PhoneService,MultipleLines,InternetService,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
5482,C00386,Female,49.0,0,No,Yes,35,Yes,Yes,Fiber optic,...,Yes,Yes,No,No,Month-to-month,Yes,Mailed check,74.44,2370.01,Yes
5405,C00386,Female,49.0,0,No,Yes,35,Yes,Yes,Fiber optic,...,Yes,Yes,No,No,Month-to-month,Yes,Mailed check,74.44,2370.01,Yes
4919,C01058,Male,69.0,1,Yes,No,33,Yes,Yes,Fiber optic,...,Yes,No,No,No,Two year,No,Credit card,59.33,1822.60,No
6322,C01058,Male,69.0,1,Yes,No,33,Yes,Yes,Fiber optic,...,Yes,No,No,No,Two year,No,Credit card,59.33,1822.60,No
5995,C02386,Male,55.0,0,Yes,No,0,Yes,Yes,DSL,...,No,No,Yes,No,Month-to-month,Yes,Credit card,56.80,49.69,No
4030,C02386,Male,55.0,0,Yes,No,0,Yes,Yes,DSL,...,No,No,Yes,No,Month-to-month,Yes,Credit card,56.80,49.69,No
6503,C02789,Female,25.0,0,No,No,30,Yes,No,Fiber optic,...,Yes,Yes,No,No,One year,Yes,Bank transfer,70.17,1955.86,No
4811,C02789,Female,25.0,0,No,No,30,Yes,No,Fiber optic,...,Yes,Yes,No,No,One year,Yes,Bank transfer,70.17,1955.86,No
946,C05251,Male,37.0,0,Yes,No,28,Yes,No,DSL,...,No,Yes,No,No,Two year,Yes,Electronic check,43.31,1107.04,No
6751,C05251,Male,37.0,0,Yes,No,28,Yes,No,DSL,...,No,Yes,No,No,Two year,Yes,Electronic check,43.31,1107.04,No


# Inspect all categorical values

In [25]:
categorical_columns = df.select_dtypes(include="object").columns

for col in categorical_columns:
    print(f"\n--- {col} ---")
    print(df[col].value_counts(dropna=False))


--- CustomerID ---
CustomerID
C06068    2
C07014    2
C06999    2
C05251    2
C06422    2
         ..
C02246    1
C02108    1
C04646    1
C03002    1
C05735    1
Name: count, Length: 7250, dtype: int64

--- Gender ---
Gender
Female    3650
Male      3612
Name: count, dtype: int64

--- Partner ---
Partner
No     4091
Yes    3142
NO       16
yes      13
Name: count, dtype: int64

--- Dependents ---
Dependents
No     5610
Yes    1623
NO       21
yes       8
Name: count, dtype: int64

--- PhoneService ---
PhoneService
Yes    6536
No      697
yes      26
NO        3
Name: count, dtype: int64

--- MultipleLines ---
MultipleLines
No                  3656
Yes                 2906
No phone service     700
Name: count, dtype: int64

--- InternetService ---
InternetService
Fiber optic    3486
DSL            2473
No             1303
Name: count, dtype: int64

--- OnlineSecurity ---
OnlineSecurity
No                     3976
Yes                    1983
No internet service    1303
Name: count, dtyp

C:\Users\ASUS  VIVOBOOK\AppData\Local\Temp\ipykernel_12892\709346951.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_columns = df.select_dtypes(include="object").columns


# Calculate missing-value summary

In [26]:
missing_count = df.isnull().sum()

missing_summary = pd.DataFrame({
    "MissingCount": missing_count,
    "MissingPercentage": (missing_count / len(df)) * 100
})

missing_summary = missing_summary[
    missing_summary["MissingCount"] > 0
].sort_values(
    "MissingPercentage",
    ascending=False
)

missing_summary

,MissingCount,MissingPercentage
TechSupport,102,1.404572
PaymentMethod,88,1.211787
Age,87,1.198017
TotalCharges,87,1.198017
MonthlyCharges,58,0.798678


## Initial Observations

* The dataset contains 7,262 records and 22 columns. The intended grain appears to be one row per customer, with `CustomerID` serving as the customer identifier.
* The dataset contains 5 numerical columns and 17 object/string columns. Of the string columns, `CustomerID` is an identifier rather than an analytical categorical variable, resulting in 16 categorical variables.
* `Age` ranges from 18 to 80 years, while `Tenure` ranges from 0 to 72 months.
* `MonthlyCharges` ranges from 18.00 to 101.20, and `TotalCharges` ranges from 0.35 to 7,084.30.
* The dataset contains missing values in `Age`, `TechSupport`, `PaymentMethod`, `MonthlyCharges`, and `TotalCharges`. The missing-value percentage ranges from approximately 0.80% to 1.40%.
* `CustomerID` contains 7,250 unique values across 7,262 records. Investigation showed that the repeated customer records are exact duplicate rows, so the duplicate copies can be addressed during the data-cleaning stage.
* In `Partner`, `Dependents`, `PhoneService`, and `PaperlessBilling`, inconsistent capitalization exists, such as `Yes`/`yes` and `No`/`NO`. These values represent the same business categories and should be standardized during data cleaning.
* `SeniorCitizen` is stored as an integer encoded with 0 and 1. Although its storage type is numerical, it should be treated analytically as a binary categorical variable.
* `Churn` contains two categories: `Yes` and `No`. There are 1,840 churned customers and 5,422 non-churned customers, giving an overall churn rate of approximately 25.34%.
* Some service-related columns contain the category `No internet service`. This represents a valid business condition and should not automatically be treated as a missing value.
* Further data cleaning and validation are required before performing detailed exploratory analysis.


In [27]:
df[numerical_columns].isnull().sum()

Age               87
SeniorCitizen      0
Tenure             0
MonthlyCharges    58
TotalCharges      87
dtype: int64

In [28]:
df["Age"].value_counts().sort_index().head(20)

Age
18.0    361
19.0     62
20.0     60
21.0     67
22.0     60
23.0     76
24.0     77
25.0     92
26.0    100
27.0    122
28.0    123
29.0    115
30.0    148
31.0    146
32.0    156
33.0    156
34.0    175
35.0    153
36.0    164
37.0    172
Name: count, dtype: int64

In [29]:
df["Tenure"].value_counts().sort_index()

Tenure
0     635
1      77
2      75
3      58
4      72
     ... 
68     25
69     14
70     15
71     16
72    132
Name: count, Length: 73, dtype: int64

In [30]:
df["SeniorCitizen"].value_counts(dropna=False)

SeniorCitizen
0    6771
1     491
Name: count, dtype: int64

In [31]:
df["MonthlyCharges"].sort_values().head(10)

1642    18.0
39      18.0
7244    18.0
6983    18.0
1785    18.0
3937    18.0
3963    18.0
1728    18.0
1647    18.0
7227    18.0
Name: MonthlyCharges, dtype: float64

In [32]:
df["TotalCharges"].sort_values().head(10)

4633    0.35
6077    0.42
6594    0.78
7016    0.79
750     1.09
2115    1.10
4487    1.31
6829    1.31
5804    2.73
411     3.94
Name: TotalCharges, dtype: float64

## Conclusion

The initial data-understanding phase established the structure, grain, variable types, distributions, and major data-quality concerns in the customer dataset.

The dataset contains 7,262 customer records and 22 columns, with `Churn` representing the primary business outcome. The overall observed churn rate is approximately 25.34%.

Several data-quality issues were identified, including missing values, exact duplicate records, and inconsistent capitalization in selected categorical variables. `SeniorCitizen` is stored as a binary integer but will be treated as a categorical variable for analysis.

The initial numerical range checks did not identify obvious invalid values. Values such as `Tenure = 0` and low `TotalCharges` values will be retained for further investigation because they may represent legitimate customer situations rather than data errors.

The identified data-quality issues will be addressed systematically in the data-cleaning stage before conducting detailed exploratory analysis.
